src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1330,6 +1331,32 @@ citus_update_node(PG_FUNCTION_ARGS)
 }
 
 
+Datum
+citus_pause_node(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	int32 nodeId = PG_GETARG_INT32(0);
+	List *placementList = NIL;
+
+	WorkerNode *workerNode = FindNodeAnyClusterByNodeId(nodeId);
+	if (workerNode == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_NO_DATA_FOUND),
+						errmsg("node %u not found", nodeId)));
+	}
+
+
+	if (NodeIsPrimary(workerNode))
+	{
+		placementList = AllShardPlacementsOnNodeGroup(workerNode->groupId);
+		LockShardsInPlacementListMetadata(placementList, AccessExclusiveLock);
```

Comment:
This method would release all the locks at the end of transaction. I mean below execution would not pause the node at all.

```sql
select citus_pause_node(2);
```

But below would pause it until transaction ends.
```sql
begin;
select citus_pause_node(2);
```

I think the method should still pause even if we are not inside a transaction. You can wait for query cancellation inside a loop just after taking the locks. If user cancels, then return. What do you think?

Reply:
citus_update_node with the begin block. I modeled that extension. @JelteF what do you think?

Reply:
If we make such a change, we need to evaluate such UDFs as well, since it may be confusing for the users; some inside transaction, some not

Reply:
> citus_update_node is also used with the begin block

I think for citus_update_node it makes sense, because it is a one time operation, you update the metadata and you are done.

For citus_pause_node, that doesn't conceptually make sense. So, for me to understand more:

- what is the use case for `citus_update_node`? Is that mostly intended for use during failovers? If so, maybe the current approach might make sense, but we should enforce that the UDF is called within a transaction block
- Again, relevant to the above question. If this is intended for use during failovers, then it is fine that we block the transactions. However, if this is a generic UDFs that users can pause traffic on a node, acquiring locks would quickly block all backends, and net effect would be all traffic to the whole cluster is blocked (e.g., there are no free backends to continue querying the rest of the cluster, and/or we'd quickly hit max_connections)
- If we require transaction blocks, than some of the uses cases of citus_pause_node would not be supported. Say, you pause traffic to node 1, but then you also need to restart the coordinator. Then, the locks would be release

Reply:
> what is the use case for citus_update_node? Is that mostly intended for use during failovers?

(I'm going to assume you meant citus_pause_node here). It's initial intended use is indeed during failover. It's basically intended to do the locking of citus_update_node, but without having to specify a new hostname+port. The main usecase  here is for Patroni, because at the start of a switchover it does not know what node will be switched over to. But during a switchover we want to pause all the traffic to that worker so that no errors occur, only a little bit of latency increase.

So right now it pauses traffic using the quite hacky `citus_update_node(123, 'some-non-existing-hostname', 5432)`

Another usecase could be controlled restarts of  a worker. Where we pause traffic to the worker on the coordinator, while the restart is happening.

I do agree that the name is quite confusing. Maybe we should call it something like `citus_pause_node_during_transaction()` instead.

Reply:
or citus_transaction_bound_pause_node()  / citus_pause_node_within_txn()

Reply:
Changing its name to  citus_pause_node_within_txn() @onderkalaci @JelteF is it ok?

Reply:
sounds good to me